### PR TITLE
Add support to 2017 Lexus is300h (AVE30R)

### PR DESCRIPTION
**Pull request by eFiniLan**
Accidentally deleted my original branch:

* Longitudinal Control by stock ACC (limit to minimum speed of 46 kph),  0:343 never show up in cabana.
* RHD, camera offset best set to 0. but default value is fine.
* 26 degree mount, camera vertically inline with FRC.
* 6 min video: https://youtu.be/bt75YNQpR5o
* Cabana: https://community.comma.ai/cabana/?route=1dd19ceed0ee2b48%7C2018-12-22--17-36-49&max=32&url=https%3A%2F%2Fchffrprivate.blob.core.windows.net%2Fchffrprivate3%2Fv2%2F1dd19ceed0ee2b48%2F6e880767f4095c51b8f2265c583313f3_2018-12-22--17-36-49

### CODE REVIEW

File: selfdrive/car/toyota/carstate.py [hunk 1]
```diff
@@ -145,28 +167,44 @@ def update(self, cp, cp_cam):
...
+      self.v_cruise_pcm = cp.vl["PCM_CRUISE_ISH"]['SET_SPEED']
+      # ish does not have curise status value (always 0), so we use curise_active value instead
+      self.pcm_acc_status = cp.vl["PCM_CRUISE"]['CRUISE_ACTIVE']
+      self.pcm_acc_active = bool(cp.vl["PCM_CRUISE"]['CRUISE_ACTIVE'])
+      self.gas_pressed = not cp.vl["PCM_CRUISE"]['GAS_RELEASED']
+      self.low_speed_lockout = False
+    else:
+      self.v_cruise_pcm = cp.vl["PCM_CRUISE_2"]['SET_SPEED']
+      self.pcm_acc_status = cp.vl["PCM_CRUISE"]['CRUISE_STATE']
+      self.pcm_acc_active = bool(cp.vl["PCM_CRUISE"]['CRUISE_ACTIVE'])
```
Comment (legonigel) [inline selfdrive/car/toyota/carstate.py]: Can common code be abstracted from the If statement?

File: selfdrive/car/toyota/carstate.py [hunk 2]
```diff
@@ -60,6 +55,33 @@ def get_can_parser(CP):
...
+      ("STEER_TORQUE_SENSOR", 50),
+      ("EPS_STATUS", 25),
+    ]
+
+    signals += [
+      ("MAIN_ON", "PCM_CRUISE_ISH", 0),
+      ("SET_SPEED", "PCM_CRUISE_ISH", 0),
+      ("AUTO_HIGH_BEAM", "LIGHT_STALK_ISH", 0),
+    ]
+
```
Comment (legonigel) [inline selfdrive/car/toyota/carstate.py]: Can you clean this code up? Something like this:
```suggestion
  checks = [
    ("WHEEL_SPEEDS", 80),
    ("STEER_ANGLE_SENSOR", 80),
    ("PCM_CRUISE", 33),
    ("STEER_TORQUE_SENSOR", 50),
    ("EPS_STATUS", 25),
  ]

  if CP.carFingerprint == CAR.LEXUS_ISH:
    checks += [
      ("BRAKE_MODULE", 50),
      ("GAS_PEDAL", 50),
      ("PCM_CRUISE_ISH", 1),
    ]
    signals += [
      ("MAIN_ON", "PCM_CRUISE_2", 0),
      ("SET_SPEED", "PCM_CRUISE_2", 0),
      ("LOW_SPEED_LOCKOUT", "PCM_CRUISE_2", 0),
      ("IPAS_STATE", "EPS_STATUS", 1),
      ("AUTO_HIGH_BEAM", "LIGHT_STALK", 0),
    ]

  else:
    checks += [
      ("BRAKE_MODULE", 40),
      ("GAS_PEDAL", 33),
      ("PCM_CRUISE_2", 33),
    ]

     signals += [
      ("MAIN_ON", "PCM_CRUISE_ISH", 0),
      ("SET_SPEED", "PCM_CRUISE_ISH", 0),
      ("AUTO_HIGH_BEAM", "LIGHT_STALK_ISH", 0),
    ]
```

Comment (eFiniLan) [reply selfdrive/car/toyota/carstate.py]: @legonigel no worries, sorted

**Comment by quillford** [comment]
so the IS has some peculiarities because the cruise stalk is wired to the DSU instead of the PCM like the supported toyotas/lexuses.

Does this cancel when the gas is pressed? I found a [cancel request](https://github.com/quillford/openpilot/blob/lexus_is_pr/opendbc/generator/toyota/_toyota_2017.dbc#L72) in the PCM_CRUISE msg which I had to use in my IS port.

I also suspect that the signal we're both using for speed is a UI msg and doesn't have a fixed unit, so before this is upstreamed this should check whether the car is set for km or miles.

**Comment by eFiniLan** [comment]
> so the IS has some peculiarities because the cruise stalk is wired to the DSU instead of the PCM like the supported toyotas/lexuses.
> 
> Does this cancel when the gas is pressed? I found a [cancel request](https://github.com/quillford/openpilot/blob/lexus_is_pr/opendbc/generator/toyota/_toyota_2017.dbc#L72) in the PCM_CRUISE msg which I had to use in my IS port.
> 
> I also suspect that the signal we're both using for speed is a UI msg and doesn't have a fixed unit, so before this is upstreamed this should check whether the car is set for km or miles.

@quillford, do you mean the OP cancelled when the gas is pressed? I don't think you need another message to detect that part as [OP handles that] (https://github.com/commaai/openpilot/blob/8b7dfb647f167d71a92ef57b07491814f7f18e9d/selfdrive/car/toyota/interface.py#L349)

the unit change is a good idea, I guess we can use UI_SEETING to detect the unit. :)

**Comment by quillford** [comment]
@eFiniLan does your IS cancel the acc when the gas is pressed? since this is a lat only port, longitudinal control is handled by the stock system. stock ACC on the IS doesn't use 0x343 but openpilot does for cancellation. hence, the use of the cancel request signal in PCM_CRUISE.

**Comment by eFiniLan** [comment]
@quillford ah I see what you are trying to do there after reading your branch, not sure if that's a good idea or not to spam 0x1d2 to cancel ACC though, OP spam 0x343 to cancel cruise control since it's for cruise control, but 0x1d2 to me has messages more than just cruise control,  I'm just afraid it will have side effects.

UI_SEETING does not tell us what unit the car is using, so need to find other ways :/ I'll leave the PR as is now since it's working perfectly.

speaking of which have you figure out a way to enable OP long control yet?

**Comment by quillford** [comment]
I have two solutions that should work but haven't had time to implement them yet. most straightforward solution is to add some hardware like a panda with firmware that would make the IS look more like the supported toyotas where set speed, main on, etc. is handled by the pcm.

File: selfdrive/car/toyota/carstate.py [hunk 3]
```diff
@@ -37,29 +37,47 @@ def get_can_parser(CP):
...
+      ("PCM_CRUISE_ISH", 1),
+    ]
+
+    signals += [
+      ("MAIN_ON", "PCM_CRUISE_ISH", 0),
+      ("SET_SPEED", "PCM_CRUISE_ISH", 0),
+      ("AUTO_HIGH_BEAM", "LIGHT_STALK_ISH", 0),
+    ]
+  else:
+    checks = [
```
Comment (legonigel) [inline selfdrive/car/toyota/carstate.py]: I think this should be +=
```suggestion
    checks += [
```

Comment (eFiniLan) [reply selfdrive/car/toyota/carstate.py]: silly mistakes, fixed :)

File: opendbc/generator/toyota/lexus_is_hybrid_2017_pt.dbc [hunk 4]
```diff
@@ -0,0 +1,41 @@
+CM_ "IMPORT _toyota_2017.dbc"
```
[the remark this answers is not in the document]
Comment (eFiniLan) [reply opendbc/generator/toyota/lexus_is_hybrid_2017_pt.dbc]: moved, https://github.com/commaai/opendbc/pull/146

File: selfdrive/car/toyota/carstate.py [hunk 5]
```diff
@@ -37,29 +37,47 @@ def get_can_parser(CP):
...
     ("WHEEL_SPEEDS", 80),
     ("STEER_ANGLE_SENSOR", 80),
     ("PCM_CRUISE", 33),
-    ("PCM_CRUISE_2", 33),
     ("STEER_TORQUE_SENSOR", 50),
     ("EPS_STATUS", 25),
   ]
 
+  if CP.carFingerprint == CAR.LEXUS_ISH:
+    checks = [
```
[the remark this answers is not in the document]
Comment (eFiniLan) [reply selfdrive/car/toyota/carstate.py]: fixed

**Comment by legonigel** [comment]
Can you please provide a cabana or explorer link for a drive using the latest commit, and a link with the system in stock mode for us to verify. Otherwise this is looking pretty good.

**Comment by eFiniLan** [comment]
> Can you please provide a cabana or explorer link for a drive using the latest commit, and a link with the system in stock mode for us to verify. Otherwise this is looking pretty good.


@legonigel thanks! 👍 

Please see the cabana log below and let me know if you need a longer records :)

Stock: https://community.comma.ai/cabana/?route=1dd19ceed0ee2b48%7C2019-03-15--12-57-37&max=14&url=https%3A%2F%2Fchffrprivate.blob.core.windows.net%2Fchffrprivate3%2Fv2%2F1dd19ceed0ee2b48%2Fdd00ded36eab4097d339fa503a99152e_2019-03-15--12-57-37

EON: https://community.comma.ai/cabana/?route=1dd19ceed0ee2b48%7C2019-03-15--12-44-40&max=12&url=https%3A%2F%2Fchffrprivate.blob.core.windows.net%2Fchffrprivate3%2Fv2%2F1dd19ceed0ee2b48%2Ff3c08c9ec2bb80312f2e4b77cba38a7a_2019-03-15--12-44-40

**Comment by rbiasini** [comment]
@legonigel is this PR ok to be merged?

**Comment by rbiasini** [comment]
@quillford @eFiniLan , now that we support the Lexus IS, I think this PR should be a lot simpler. Right?

**Comment by eFiniLan** [comment]
@rbiasini , thanks for getting it supported finally, ISH still a bit different to the IS I believe, I have to temporary set 

```checks = []```

in carvalues.py otherwise I'm getting commIssue error, I guess we will still need:
```
  if CP.carFingerprint == CAR.LEXUS_ISH:
    checks += [
      ("BRAKE_MODULE", 50),
      ("GAS_PEDAL", 50),
    ]
```
I'll do more testing and create a separate PR if required : )

**Comment by dragonpilot** [comment]
@rbiasini, just finished testing, we still need several changes to make it to work:
1. ISH need PCM_CRUISE_ALT instead of DSU_CRUISE.
2. ISH's GAS_PEDAL is on 581 (0x245) like most of the toyota, not 705 (0x2C1).
3. ISH's ```cp.vl["PCM_CRUISE"]['CRUISE_STATE']``` is always 0.
4. a few message is running different speed, [see this report](https://github.com/commaai/openpilot/files/3603801/lexus_is_2018_pt_generated-1568260419779.csv.report.txt):
```
      ("BRAKE_MODULE", 50),
      ("GAS_PEDAL_ALT", 50),
      ("PCM_CRUISE_ALT", 1),
``` 
[cabana log](https://my.comma.ai/cabana/?route=1dd19ceed0ee2b48%7C2019-09-08--16-08-22&url=https%3A%2F%2Fchffrprivate-vzn.azureedge.net%2Fchffrprivate3%2Fv2%2F1dd19ceed0ee2b48%2F24d2c2a6a60a60229f5c6b713016a637_2019-09-08--16-08-22&seekTime=2)

**Comment by eFiniLan** [comment]
@rbiasini, just finished testing, we still need several changes to make it to work:
1. ISH need PCM_CRUISE_ALT instead of DSU_CRUISE.
2. ISH's GAS_PEDAL is on 581 (0x245) like most of the toyota, not 705 (0x2C1).
3. ISH's ```cp.vl["PCM_CRUISE"]['CRUISE_STATE']``` is always 0.
4. a few message is running different speed, [see this report](https://github.com/commaai/openpilot/files/3603801/lexus_is_2018_pt_generated-1568260419779.csv.report.txt):
```
      ("BRAKE_MODULE", 50),
      ("GAS_PEDAL_ALT", 50),
      ("PCM_CRUISE_ALT", 1),
``` 
[cabana log](https://my.comma.ai/cabana/?route=1dd19ceed0ee2b48%7C2019-09-08--16-08-22&url=https%3A%2F%2Fchffrprivate-vzn.azureedge.net%2Fchffrprivate3%2Fv2%2F1dd19ceed0ee2b48%2F24d2c2a6a60a60229f5c6b713016a637_2019-09-08--16-08-22&seekTime=2)

**Comment by geohot** [comment]
If this pull request is still relevant, please resubmit it to the master branch. This should allow our development flow to work better, we commit to master, you commit to master, everyone commits to master! Learn more: https://medium.com/@comma_ai/a-2020-theme-externalization-13b33326d8b3

**Pull request closed without merging**